src/bid/encoding.rs

```diff
@@ -62,72 +68,70 @@ impl Into<StorageScalar> for Bid {
-        let mut messages: Vec<Variable> = vec![];
-        messages.push(composer.add_input(type_fields));
-        // Push both JubJubAffine coordinates as a Scalar.
-        self.encrypted_data.cipher().iter().for_each(|c| {
-            let c = composer.add_input(*c);
-            messages.push(c);
-        });
-
-        // Push both JubJubAffine coordinates as a Scalar.
-        messages.push(composer.add_input(self.stealth_address.pk_r().get_x()));
-        messages.push(composer.add_input(self.stealth_address.pk_r().get_y()));
-        // Push both JubJubAffine coordinates as a Scalar.
-        messages.push(composer.add_input(self.stealth_address.R().get_x()));
-        messages.push(composer.add_input(self.stealth_address.R().get_y()));
-        messages.push(composer.add_input(self.hashed_secret));
-        // Push both JubJubAffine coordinates as a Scalar.
-        messages.push(composer.add_input(self.c.get_x()));
-        messages.push(composer.add_input(self.c.get_y()));
-        // Add elebility & expiration timestamps.
-        messages.push(composer.add_input(self.elegibility_ts));
-        messages.push(composer.add_input(self.expiration_ts));
-
-        // Perform the sponge_hash inside of the Constraint System
-        let storage_bid_hash = sponge_hash_gadget(composer, &messages);
-        // Constraint the hash to be equal to the real one
-        let real_hash: StorageScalar = self.clone().into();
-        let real_hash: BlsScalar = real_hash.into();
-        composer.constrain_to_constant(
-            storage_bid_hash,
-            BlsScalar::zero(),
-            -real_hash,
-        );
-        storage_bid_hash
+impl Into<StorageScalar> for Bid {
+    fn into(self) -> StorageScalar {
+        (&self).into()
     }
 }
 
+/// Hashes the internal Bid parameters using the Poseidon252 hash
+/// function and the cannonical encoding for hashing returning a
+/// Variable which contains the hash of the Bid.
+pub(crate) fn preimage_gadget(
+    composer: &mut StandardComposer,
+    encrypted_data: (Variable, Variable),
+    commitment: PlonkPoint,
+    // (Pkr, R)
+    stealth_addr: (PlonkPoint, PlonkPoint),
+    hashed_secret: Variable,
+    elegibility_ts: Variable,
+    expiration_ts: Variable,
+) -> Variable {
+    // This field represents the types of the inputs and has to be the same
+    // as the default one.
+    // It has been already checked that it's safe to unwrap here since the
+    // value fits correctly in a `BlsScalar`.
+    let type_fields = BlsScalar::from_bytes(&TYPE_FIELDS).unwrap();
+
+    // Add to the composer the values required for the preimage.
+    let mut messages: Vec<Variable> = vec![];
+    messages.push(composer.add_input(type_fields));
+    // Push cipher as scalars.
+    messages.push(encrypted_data.0);
+    messages.push(encrypted_data.1);
+
+    // Push both JubJubAffine coordinates as a Scalar.
+    messages.push(*stealth_addr.0.x());
```

Comment:
Also it looks odd that we have a `to_hash_inputs` for the points but we don't have a counterpart for the circuit. Maybe also leave a TODO note to create one in Plonk.

Reply:
hmmm. I'm not sure. The point of `to_hash_inputs` was to avoid people hasing Extended Coordinates. In plonk we work with AffinePoints always. At least the types are always exposed like them. Therefore IDK if it should apply.